text/0000-inline-semantic.md

```diff
@@ -0,0 +1,989 @@
+let g: fn(Option<u32>) -> u32 = Option::unwrap;
+assert!(f == g); // This must remain `true`.
+f(None);
+g(None); // The effect of these two calls must be the same.
+```
+
+## Standard libraries
+
+The `caller_location()` intrinsic returns the `Location` structure which encodes the file, line and
+column of the callsite. This shares the same structure as the existing type `std::panic::Location`.
+Therefore, the type is promoted to a lang-item, and moved into `core::panicking::Location`. It is
+re-exported in `libstd`.
+
+Thanks to how `#[implicit_caller_location]` is implemented, we could provide a safe wrapper around
+the `caller_location()` intrinsic:
+
+```rust
+impl<'a> Location<'a> {
+    #[implicit_caller_location]
+    pub fn caller() -> Location<'static> {
+        unsafe {
+            ::intrinsics::caller_location()
+        }
+    }
+}
+```
+
+The `panic!` macro is modified to use `Location::caller()` (or the intrinsic directly) so it can
+report the caller location inside `#[implicit_caller_location]`.
+
+```rust
+macro_rules! panic {
+    ($msg:expr) => {
+        let loc = $crate::panicking::Location::caller();
+        $crate::panicking::panic(&($msg, loc.file(), loc.line(), loc.column()))
+    };
+    ...
+}
+```
+
+Actually this is now more natural for `core::panicking::panic_fmt` to take `Location` directly
+instead of tuples, so one should consider changing their signature. But this is out-of-scope for
+this RFC.
+
+`panic!` is often used about of `#[implicit_caller_location]` functions. In those cases, the
+`caller_location()` intrinsic will pass unchanged through all MIR passes into trans. As a fallback,
+the intrinsic will expand to `Location { file: file!(), line: line!(), col: column!() }` during
+trans.
+
+## Runtime-free backtrace for `?` operator
+
+The standard `Try` implementations could participate in specialization, so external crates like
+`error_chain` could provide a specialized impl that prepends the caller location into the callstack
+everytime `?` is used.
+
+```rust
+// libcore:
+impl<T, E> Try for Result<T, E> {
+    type Ok = T;
+    type Error = E;
+
+    fn into_result(self) -> Self { self }
+    fn from_ok(v: Self::Ok) -> Self { Ok(v) }
+    default fn from_error(e: Self::Error) -> Self { Err(e) }
+//  ^~~~~~~
+}
+
+// my_crate:
+impl<T> Try for Result<T, my_crate::error::Error> {
+    #[implicit_caller_location]
+    fn from_error(mut e: Self::Error) -> Self {
+        e.call_stack.push(Location::caller());
+        Err(e)
+    }
+}
+```
+
+## Location detail control
+
+An unstable flag `-Z location-detail` is added to `rustc` to control how much factural detail will
+be emitted when using `caller_location()`. User can toggle `file`, `line` and `column` separately,
+e.g. when compiling with:
+
+```sh
+rustc -Zlocation-detail=line
+```
+
+only the line number will be real, and the file and column will always be dummy value like
+
+    thread 'main' panicked at 'error message', <redacted>:192:0
```

Comment:
Being able to chop off the path at the crate level may also be useful. For example, if I have an error in `src/lib.rs`, Cargo should be able to tell rustc to make this `<cratename:crateversion>/src/lib.rs`. Probably flags like `-Z location-root=/path/to/folder -Zlocation-prefix=<cratename:crateversion>`

These flags will probably also be useful elsewhere in the compiler (for debuginfo for example), and may improve prospects for deterministic builds.